src/SlamData/Halogen/Select/Cascade/Component.purs

```diff
@@ -0,0 +1,229 @@
+cascadeFormClass = className "cascade-form"
+
+cascadeNewSelectClass :: ClassName
+cascadeNewSelectClass = className "cascade-new-select"
+
+cascadeSelectClass :: ClassName
+cascadeSelectClass = className "cascade-select"
+
+newtype SelectKey = SelectKey String
+runSelectKey :: SelectKey -> String
+runSelectKey (SelectKey s) = s
+
+instance eqSelectKey :: Eq SelectKey where
+  eq (SelectKey s) (SelectKey s') = s == s'
+
+instance ordSelectKey :: Ord SelectKey where
+  compare (SelectKey s) (SelectKey s') = compare s s'
+
+instance showSelectkey :: Show SelectKey where
+  show (SelectKey s) = "(SelectKey " <> s <> ")"
+
+newtype Option v =
+  Option { key :: SelectKey
+         , value :: v
+         }
+
+runOption :: forall v. Option v -> { key :: SelectKey, value :: v }
+runOption (Option r) = r
+
+instance eqOption :: Eq (Option v) where
+  eq (Option {key = k}) (Option {key = k'}) = k == k'
+
+instance ordOption :: Ord (Option v) where
+  compare (Option {key = k}) (Option {key = k'}) = compare k k'
+
+instance showOption :: (Show v) => Show (Option v) where
+  show (Option {key, value}) =
+    "Option {key = " <> show key <> ", value = " <> show value <> "}"
+
+type State v =
+  {
+    options :: Set.Set (Option v)
+  , selected :: Map.Map Int (Option v)
+  , keyMap :: Map.Map SelectKey (Option v)
+  }
+
+_options :: forall a r. LensP {options :: a|r} a
+_options = lens _.options _{options = _}
+
+_selected :: forall a r. LensP {selected :: a|r} a
+_selected = lens _.selected _{selected = _}
+
+initialState :: forall v. Array v -> (v -> SelectKey) -> State v
+initialState vs mkKey =
+  {
+    selected: Map.empty
+  , options: Set.fromFoldable opts
+  , keyMap: Map.fromFoldable $ map optToTuple opts
+  }
+  where
+  mkOpt :: v -> Option v
+  mkOpt v = Option {value: v, key: mkKey v}
+
+  opts :: Array (Option v)
+  opts = map mkOpt vs
+
+  optToTuple :: Option v -> Tpl.Tuple SelectKey (Option v)
+  optToTuple opt@(Option {key}) = Tpl.Tuple key opt
+
+data Query v a
+  = Selected Int String a
+  | Remove Int a
+  | GetSelected (Array v -> a)
+
+type CascadeSelectDSL v e = ComponentDSL (State v) (Query v) (Aff e)
+
+type CascadeSelectConfig =
+  {
+    inviteMessage :: M.Maybe String
+  }
+
+cascadeSelect
+  :: forall v e
+   . CascadeSelectConfig
+  -> Component (State v) (Query v) (Aff e)
+cascadeSelect cfg = component (render cfg) eval
+
+type RenderAccum v =
+  {
+    selectable :: Set.Set (Option v)
+  , html :: Array (ComponentHTML (Query v))
+  }
+
+initialAccum :: forall v. State v -> RenderAccum v
+initialAccum state =
+  {
+    selectable: state.options
+  , html: [ ]
+  }
+
+
+render
+  :: forall v
+   . CascadeSelectConfig
+  -> State v
+  -> ComponentHTML (Query v)
+render cfg state =
+  H.form [ P.classes [ cascadeFormClass ] ]
+    (
+      htmlAndSelectable.html
+      <> newSelect
+    )
+  where
+  inviteMessage :: String
+  inviteMessage =
+    M.fromMaybe "Please, select something" cfg.inviteMessage
```

Comment:
Perhaps we should avoid making this message a `Maybe` - I don't think there's a sensible default to fall back on, given the slightly strange interaction this component provides.